pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,370 @@
+}
+
+const (
+	downwardAPIPluginName = "kubernetes.io/downward-api"
+)
+
+// downwardAPIPlugin implements the VolumePlugin interface.
+type downwardAPIPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &downwardAPIPlugin{}
+
+func (plugin *downwardAPIPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *downwardAPIPlugin) Name() string {
+	return downwardAPIPluginName
+}
+
+func (plugin *downwardAPIPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.VolumeSource.DownwardAPI != nil
+}
+
+func (plugin *downwardAPIPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	v := &downwardAPIVolume{
+		volName: spec.Name,
+		pod:     pod,
+		podUID:  pod.UID,
+		plugin:  plugin,
+		mounter: mounter}
+	v.fieldReferenceFileNames = make(map[string]string)
+	for _, fileInfo := range spec.VolumeSource.DownwardAPI.Items {
+		v.fieldReferenceFileNames[fileInfo.FieldRef.FieldPath] = path.Clean(fileInfo.Path)
+	}
+	return &downwardAPIVolumeBuilder{
+		downwardAPIVolume: v,
+		opts:              &opts}, nil
+}
+
+func (plugin *downwardAPIPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &downwardAPIVolumeCleaner{&downwardAPIVolume{volName: volName, podUID: podUID, plugin: plugin, mounter: mounter}}, nil
+}
+
+// downwardAPIVolume retrieves downward API data and placing them into the volume on the host.
+type downwardAPIVolume struct {
+	volName                 string
+	fieldReferenceFileNames map[string]string
+	pod                     *api.Pod
+	podUID                  types.UID // TODO: remove this redundancy as soon NewCleaner func will have *api.POD and not only types.UID
+	plugin                  *downwardAPIPlugin
+	mounter                 mount.Interface
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// downwardAPIVolumeBuilder fetches info from downward API from the pod
+// and dumps it in files
+type downwardAPIVolumeBuilder struct {
+	*downwardAPIVolume
+	opts *volume.VolumeOptions
+}
+
+// downwardAPIVolumeBuilder implements volume.Builder interface
+var _ volume.Builder = &downwardAPIVolumeBuilder{}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+// The internal sync interval of kubelet will drive the refresh of data.
+// TODO: Add volume specific ticker and refresh loop
+func (b *downwardAPIVolumeBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+func (b *downwardAPIVolumeBuilder) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a downwardAPI volume %v for pod %v/%v at %v", b.volName, b.pod.Namespace, b.pod.Name, dir)
+	// Wrap EmptyDir,let it do the setup.
```

Comment:
Comment that this relies on the idempotency of the wrapped plugin to avoid repeatedly mounting.